libs/agent-sm/agent/src/main/java/org/opensearch/javaagent/Agent.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.javaagent;
+
+import org.opensearch.javaagent.bootstrap.AgentPolicy;
+
+import java.lang.instrument.Instrumentation;
+import java.nio.channels.SocketChannel;
+import java.util.Map;
+
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.ClassFileLocator;
+import net.bytebuddy.dynamic.loading.ClassInjector;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.matcher.ElementMatcher.Junction;
+import net.bytebuddy.matcher.ElementMatchers;
+
+/**
+ * Java Agent
+ */
+public class Agent {
+    /**
+     * Constructor
+     */
+    private Agent() {}
+
+    /**
+     * Premain
+     * @param agentArguments agent arguments
+     * @param instrumentation instrumentation
+     * @throws Exception Exception
+     */
+    public static void premain(String agentArguments, Instrumentation instrumentation) throws Exception {
+        initAgent(instrumentation);
+    }
+
+    /**
+     * Agent Main
+     * @param agentArguments agent arguments
+     * @param instrumentation instrumentation
+     * @throws Exception Exception
+     */
+    public static void agentmain(String agentArguments, Instrumentation instrumentation) throws Exception {
+        initAgent(instrumentation);
+    }
+
+    private static AgentBuilder createAgentBuilder(Instrumentation inst) throws Exception {
+        final Junction<TypeDescription> systemType = ElementMatchers.isSubTypeOf(SocketChannel.class);
+
+        final AgentBuilder.Transformer transformer = (b, typeDescription, classLoader, module, pd) -> b.visit(
+            Advice.to(SocketChannelInterceptor.class)
+                .on(ElementMatchers.named("connect").and(ElementMatchers.not(ElementMatchers.isAbstract())))
```

Comment:
I'm just trying to understand how this instrumentation works. It looks like this will intercept calls from a combination of the `systemType` (in this case SocketChannel) and the method `connect` that is not an abstract method? If the call meets these conditions then `SocketChannelInterceptor.intercept` is called?

Reply:
I see that we also mention the action `resolve` when creating the SocketPermission in the interceptor.

```
final SocketPermission permission = new SocketPermission(host, "connect,resolve");
```

Is `resolve` required when instantiating the SocketPermission even if its intercepting the `connect` call?

Reply:
> Is `resolve` required when instantiating the SocketPermission even if its intercepting the `connect` call?

This is DNS call, we usually grant both in all places (since not all connect calls have IP address)

Reply:
> I'm just trying to understand how this instrumentation works

We only retransform classes that implement `connect` method (SocketChannel::connect is abstract), we don't touch any other classes in the hieararchy